syntax = "proto3";

option csharp_namespace = "ExcelMinder.Shared";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

service StockSimulator {
  // Returns a list of available stocks
  rpc ListStocks(google.protobuf.Empty) returns (StockList) {}

  // Returns the current price of a stock
  rpc GetStockPrice(StockRequest) returns (StockPrice) {}

  // Executes a trade on the stock market
  rpc ExecuteTrade(TradeRequest) returns (TradeResponse) {}

  rpc LastTradeRequests(PageRequest) returns (TradeRequestList) {}
  
  rpc LastExecutedTrades(PageRequest) returns (TradeResponseList) {}
  
  
  // TODO: Move to a separate service
  
}

message StockList {
  repeated Stock stocks = 1;
}

message Stock {
  string symbol = 1;
  string name = 2;
}

message StockRequest {
  string symbol = 1;
}

message StockPrice {
  string symbol = 1;
  float price = 2;
  google.protobuf.Timestamp timestamp = 3;
}

message TradeRequest {
  string symbol = 1;
  int64 quantity = 2;
  TradeType type = 3;
  float price = 4;
}

message TradeResponse {
  string symbol = 1;
  int64 quantity = 2;
  TradeType type = 3;
  float price = 4;
  float total_cost = 5;
  google.protobuf.Timestamp timestamp = 6;
}

enum TradeType {
  BUY = 0;
  SELL = 1;
}

message PageRequest {
  int32 page = 1;
  int32 page_size = 2;
}

message TradeRequestList {
  repeated TradeRequest trades = 1;
}

message TradeResponseList {
  repeated TradeResponse trades = 1;
}

// TODO: Move to a separate proto file
// External APIs (derived from Alpha Vantage messages

message RealtimeCurrencyExchangeRate {
  string from_currency_code = 1;
  string from_currency_name = 2;
  string to_currency_code = 3;
  string to_currency_name = 4;
  double exchange_rate = 5;
  string last_refreshed = 6;
  string time_zone = 7;
  double bid_price = 8;
  double ask_price = 9;
}

// TODO: Move to a separate proto file
// Excel communication messages

message CellProperties {
  string cell_value = 1;
  double text_color = 2;
  double background_color = 3;
  string font = 4;
  bool font_style = 5;
  double cell_weight = 6;
  double cell_height = 7;
  double border_color = 8;
  double border_thickness = 9;
  string border_style = 10;
}

message ExcelReport {
  string report_id = 1;
  repeated CellProperties cell_properties = 2;
} 